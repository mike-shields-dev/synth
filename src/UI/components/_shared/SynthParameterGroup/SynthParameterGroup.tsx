import { useRef, ReactNode } from "react";
import css from './SynthParameterGroup.module.css';
import { camelCaseToTitleCase } from "../../../../utils/camelCaseToTitleCase";

interface Props {
  children: (isFocused: boolean) => ReactNode;
  group: string;
  isFocused: boolean;
  updateFocus: (id: string) => void;
}

function SynthParameterGroup({
  children,
  group,
  isFocused,
  updateFocus,
}: Props) {
  const formRef = useRef<HTMLFormElement>(null);

  return (
    <section
      id={group}
      className={`
        ${css.SynthParameterGroup}
        ${css[`SynthParameterGroup${isFocused ? "--focus" : ""}`]}`
      }
        aria-label={camelCaseToTitleCase(group)}
        onFocus={(e) => updateFocus(e.currentTarget.id)}
        onClick={(e => updateFocus(e.currentTarget.id))}
        ref={formRef}
      >
        <header>
          <h2>{camelCaseToTitleCase(group)}</h2>
      </header>
      <div className={css.SynthParameterGroup__controls}>
        { children && children(isFocused) }
      </div>
      </section>
  )
}

export default SynthParameterGroup;
  