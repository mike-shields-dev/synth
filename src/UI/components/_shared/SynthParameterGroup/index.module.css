.SynthParameterGroup {
    display: grid;
    gap: 1rem;
    border-radius: 4px;
    padding: 0.5rem 1rem;
    background: rgba(0, 0, 0, 0.125);
    color: #c9c9c9;
    transition: background 0.5s;
    width: 100%;
    height: 100%;
}

.SynthParameterGroup--focus {
    background: rgba(255, 255, 255, 0.125);
    color: #fff;
}

.SynthParameterGroup__controls {
    display: flex;
    justify-content: space-around;
}
