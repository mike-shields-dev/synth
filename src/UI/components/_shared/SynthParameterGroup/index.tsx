import { useRef, ReactNode } from "react";
import css from './index.module.css';
import { camelCaseToTitleCase } from "../../../../utils/camelCaseToTitleCase";

interface Props {
  children: (isFocused: boolean) => ReactNode;
  group: string;
  isFocused: boolean;
  updateFocus: (id: string) => void;
}

function SynthParameterGroup({
  children,
  group,
  isFocused,
  updateFocus,
}: Props) {
  const formRef = useRef<HTMLFormElement>(null);

  return (
    <section
      className={`
        ${css.SynthParameterGroup}
        ${css[`SynthParameterGroup${isFocused ? "--focus" : ""}`]}`
      }
        aria-label={camelCaseToTitleCase(group)}
        id={group}
        onFocus={(e) => updateFocus(e.currentTarget.id)}
        onClick={(e => updateFocus(e.currentTarget.id))}
        ref={formRef}
      >
        <header>
          <h2>{camelCaseToTitleCase(group)}</h2>
        </header>
        { children && children(isFocused) }
      </section>
  )
}

export { SynthParameterGroup };
  