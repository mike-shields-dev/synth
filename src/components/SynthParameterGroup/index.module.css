.SynthParameterGroup > form:focus-within {
    border: 1px solid red;
}