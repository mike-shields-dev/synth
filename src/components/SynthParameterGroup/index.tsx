import React from "react";
import css from './index.module.css';

interface Props {
    groupName: string;
    children: React.ReactNode;
    handleFocus: React.FocusEventHandler;
}

function SynthParameterGroup({ groupName, handleFocus, children }: Props) {
  return (
    <section className={css.SynthParameterGroup}>
      <form
        aria-label={groupName}
        id={groupName}
        onFocus={handleFocus}
      >
          { children }
      </form>
    </section>
  )
}

export { SynthParameterGroup };
