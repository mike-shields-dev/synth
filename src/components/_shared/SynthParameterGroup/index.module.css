.SynthParameterGroup > form:focus-within, 
.SynthParameterGroup--focus {
    border: 1px solid red;
}

