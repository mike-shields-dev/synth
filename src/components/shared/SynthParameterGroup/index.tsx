import { useRef } from "react";
import css from './index.module.css';

interface Props {
  children: (isFocused: boolean) => JSX.Element | JSX.Element[];
  group: string;
  title: string;
  isFocused: boolean;
  updateFocus: (id: string) => void;
}

function SynthParameterGroup({
  children = (isFocused: boolean) => <></>,
  title: headerName,
  group,
  isFocused,
  updateFocus,
}: Props) {
  const formRef = useRef<HTMLFormElement>(null);

  return (
    <section>
      <form
        className={
          css[`SynthParameterGroup${
            isFocused ? "--focus" : ""
          }`
        ]}
        aria-label={group}
        id={group}
        onFocus={(e) => updateFocus(e.currentTarget.id)}
        onClick={(e => updateFocus(e.currentTarget.id))}
        ref={formRef}
      >
        <header>
          <h2>{headerName}</h2>
        </header>
        { children && children(isFocused) }
      </form>
    </section>
  )
}

export { SynthParameterGroup };
