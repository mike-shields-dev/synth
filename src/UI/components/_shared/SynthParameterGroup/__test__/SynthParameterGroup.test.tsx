import { describe, it, expect, vi } from 'vitest';
import { fireEvent, render, screen, within } from "@testing-library/react";
import SynthParameterGroup from '../SynthParameterGroup';

const validProps = {
    group: 'Group Name',
    updateFocus: vi.fn(),
    isFocused: true,
}

const ValidChild = () => <div>valid child</div>;


describe('SynthParameterGroup', () => {
    beforeEach(() => render(
        <SynthParameterGroup {...validProps}>
            { (isFocused) => <ValidChild /> }
        </SynthParameterGroup>)
    );

    afterEach(() => vi.restoreAllMocks());

    it('matches snapshot', () => {
        const { container } = render(
            <SynthParameterGroup {...validProps}>
                { (isFocused) => <ValidChild /> }
            </SynthParameterGroup>
        );

        expect(container).toMatchSnapshot();
    });
    
    it(`renders a form with the name groupName`, () => {
        expect(
            screen.getByRole('region', { name: validProps.group })
        ).toBeInTheDocument();
    });

    it('renders a heading within the form with the given header name', () => {
        const header = screen.getByRole('heading');

        expect(screen.getByRole('heading')).toBeInTheDocument();
        expect(within(header).getByText(validProps.group)).toBeInTheDocument();
    });

    it('renders children', () => {
        expect(
            screen.getByText(/valid child/i)
        ).toBeInTheDocument();
    });

    it('invokes updateFocus callback, when a child component is focused', () => {
        const child = screen.getByText(validProps.group);

        expect(validProps.updateFocus).toHaveBeenCalledTimes(0);
        
        fireEvent.focus(child);

        expect(validProps.updateFocus).toHaveBeenCalledTimes(1);
        expect(validProps.updateFocus).toHaveBeenCalledWith(validProps.group);
    });

    it('invokes updateFocus when the user clicks anywhere within the component', () => {
        const formElement = screen.getByRole('region', { name: validProps.group });

        expect(validProps.updateFocus).toHaveBeenCalledTimes(0);
        
        fireEvent.click(formElement);

        expect(validProps.updateFocus).toHaveBeenCalledTimes(1);
        expect(validProps.updateFocus).toHaveBeenCalledWith(validProps.group);
    });
});