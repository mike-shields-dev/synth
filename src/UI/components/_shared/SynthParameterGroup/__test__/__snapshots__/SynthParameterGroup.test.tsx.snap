// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SynthParameterGroup > matches snapshot 1`] = `
<div>
  <section
    aria-label="Group  Name"
    class="
        _SynthParameterGroup_541118
        _SynthParameterGroup--focus_541118"
    id="Group Name"
  >
    <header>
      <h2>
        Group  Name
      </h2>
    </header>
    <div
      class="_SynthParameterGroup__controls_541118"
    >
      <div>
        valid child
      </div>
    </div>
  </section>
</div>
`;
